btb/session.py

```diff
@@ -0,0 +1,278 @@
+        errors = self.errors[tunable_name]
+
+        if errors >= self.max_errors:
+            LOGGER.warning('Too many errors: %s. Removing tunable %s', errors, tunable_name)
+            self._normalized_scores.pop(tunable_name, None)
+            self._tunable_names.remove(tunable_name)
+
+    def record(self, tunable_name, config, score):
+        """Record the configuration and the associated score to it inside the tuner.
+
+        Records the associated configuration and score to the tuner. Evaluates if the
+        score is the best score found, if so, updates ``self._best_normalized`` and
+        ``self.best_proposal`` with the associated to them values.
+
+        Args:
+            tunable_name (str):
+                The name of the tunable that the score and config pertain.
+            config (dict):
+                Dictionary representation of the configuration given to the tunable to obtain
+                the score.
+            score (float):
+                Obtained score with the given configuration.
+        """
+        proposal_id = self._make_id(tunable_name, config)
+        proposal = self.proposals[proposal_id]
+        proposal['score'] = score
+
+        if score is None:
+            self.handle_error(tunable_name)
+        else:
+            normalized = self._normalize(score)
+            if normalized > self._best_normalized:
+                LOGGER.info('New optimal found: %s - %s', tunable_name, score)
+                self.best_proposal = proposal
+                self._best_normalized = normalized
+            try:
+                tuner = self._tuners[tunable_name]
+                tuner.record(config, normalized)
+                self._normalized_scores[tunable_name].append(normalized)
+            except Exception:
+                LOGGER.exception('Could not record score to tuner.')
+
+    def run(self, iterations=None):
+        """Execute proposal iterations over the tunables.
+
+        Given a number of ``iterations`` propose new configurations for the tunables,
+        score and record those inside the ``tuner`` object.
+
+        Returns:
+            best_proposal (dict):
+                Best configuration found with the name of the tunable and the hyperparameters
+                and crossvalidated score obtained for it.
+        """
+        if iterations is None:
+            iterator = itertools.count()
+        else:
+            iterator = self._range(iterations)
+
+        for _ in iterator:
+            self.iterations += 1
+            proposal = self.propose()
+            if proposal is None:
+                continue
+
+            tunable_name, config = proposal
+
+            try:
+                LOGGER.debug('Scoring proposal %s - %s: %s', self.iterations, tunable_name, config)
+                score = self.scorer(tunable_name, config)
+
+            except Exception:
+                LOGGER.exception('Proposal %s - %s crashed', self.iterations, tunable_name)
```

Comment:
Also report the configuration that provoked the exception (or the proposal `id`, so it can be retrieved later on)